bin/q.py

```diff
@@ -3297,106 +3298,107 @@ def initialize_command_line_parser(p, qrc_filename):
-    output_data_option_group.add_option("-E", "--output-encoding", dest="output_encoding",
+    output_data_option_group.add_argument("-E", "--output-encoding", dest="output_encoding",
                                         default=default_output_encoding,
                                         help="Output encoding. Defaults to 'none', leading to selecting the system/terminal encoding")
-    output_data_option_group.add_option("-W", "--output-quoting-mode", dest="output_quoting_mode",
+    output_data_option_group.add_argument("-W", "--output-quoting-mode", dest="output_quoting_mode",
                                         default=default_output_quoting_mode,
                                         help="Output quoting mode. Possible values are all, minimal, nonnumeric and none. Note the slightly misleading parameter name, and see the matching -w parameter for input quoting.")
-    output_data_option_group.add_option("-L", "--list-user-functions", dest="list_user_functions",
+    output_data_option_group.add_argument("-L", "--list-user-functions", dest="list_user_functions",
                                         default=default_list_user_functions, action="store_true",
                                         help="List all user functions")
-    parser.add_option("", "--overwrite-qsql", dest="overwrite_qsql", default=default_overwrite_qsql,
+    parser.add_argument("--overwrite-qsql", dest="overwrite_qsql", default=default_overwrite_qsql,
                       help="When used, qsql files (both caches and store-to-db) will be overwritten if they already exist. Use with care.")
-    parser.add_option_group(output_data_option_group)
+    parser.add_argument_group(output_data_option_group)
     # -----------------------------------------------
-    query_option_group = OptionGroup(parser, "Query Related Options")
-    query_option_group.add_option("-q", "--query-filename", dest="query_filename", default=default_query_filename,
+    query_option_group = parser.add_argument_group("Query Related Options")
+    query_option_group.add_argument("-q", "--query-filename", dest="query_filename", default=default_query_filename,
                                   help="Read query from the provided filename instead of the command line, possibly using the provided query encoding (using -Q).")
-    query_option_group.add_option("-Q", "--query-encoding", dest="query_encoding", default=default_query_encoding,
+    query_option_group.add_argument("-Q", "--query-encoding", dest="query_encoding", default=default_query_encoding,
                                   help="query text encoding. Experimental. Please send your feedback on this")
-    parser.add_option_group(query_option_group)
+    parser.add_argument_group(query_option_group)
     # -----------------------------------------------
-    (options, args) = parser.parse_args()
-    return args, options, parser
+    parser.add_argument('leftover', nargs='*')
```

Comment:
`argparse` is almost compatibile with `optparse` but not quite; this is one of the differences. `argparse` must know if some positional arguments follow the options, while `optparse` just returned both as a tuple.
Therefore I had to improvise as per the python docs (https://docs.python.org/3/library/argparse.html):
```
Replace (options, args) = parser.parse_args() with args = parser.parse_args() and add additional ArgumentParser.add_argument() calls for the positional arguments. Keep in mind that what was previously called options, now in the argparse context is called args.

```

and named all the following positional arguments `leftover`.

Reply:
The "leftover" parameter is probably needed, as you're describing here. 

However, `dump-defaults` is writing its output so users will be able to write it directly into the `.qrc` file, so this internal parameter needs to be removed from the output, otherwise this will mislead the users.

Fixing this in the code will also allow reverting the dump-defaults test to its original form (34).